### PR TITLE
minimal working code and tests for bosch-hale reactivities

### PR DESCRIPTION
**Pull request by tmarkj**
This is part of the work for part 2.

### CODE REVIEW

**Review by RemDelaporteMathurin — COMMENTED** [review]
Can you give more details about what this code does?

**Comment by RemDelaporteMathurin** [comment]
Don't merge this before review tomorrow

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import numpy as np
+import sys
+import matplotlib.pyplot as plt
+
+## Class for Bosch-Hale reactivity
+## All of it comes from: H.-S. Bosch and G.M. Hale 1992 Nucl. Fusion 32 611
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:6]: Should be docstrings of the class with triple quotes

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+class Reactivity:
+
+	known_reactions = ["DDn", "DDp", "DTn", "D3Hep"]
+
+	# The space is here so the indexing matches the paper
+	DDn_Cs   = [" ", 5.43360e-12, 5.85778e-3, 7.68222e-3, 0.0, -2.96400e-6, 0.0, 0.0]
+	DDp_Cs   = [" ", 5.65718e-12, 3.41267e-3, 1.99167e-3, 0.0, 1.05060e-5, 0.0, 0.0]
+	DTn_Cs   = [" ", 1.1302e-9, 1.51361e-2, 7.51886e-2, 4.60643e-3, 1.35e-2, -1.0675e-4, 1.366e-5]
+	D3Hep_Cs = [" ", 5.51036e-10, 6.41918e-3, -2.02896e-3, -1.91080e-5, 1.35776e-4, 0.0, 0.0 ]
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:15]: Wrong indentation

Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:15]:
```suggestion
	known_reactions = ["DDn", "DDp", "DTn", "D3Hep"]

	# The space is here so the indexing matches the paper
	DDn_Cs   = [" ", 5.43360e-12, 5.85778e-3, 7.68222e-3, 0.0, -2.96400e-6, 0.0, 0.0]
	DDp_Cs   = [" ", 5.65718e-12, 3.41267e-3, 1.99167e-3, 0.0, 1.05060e-5, 0.0, 0.0]
	DTn_Cs   = [" ", 1.1302e-9, 1.51361e-2, 7.51886e-2, 4.60643e-3, 1.35e-2, -1.0675e-4, 1.366e-5]
	D3Hep_Cs = [" ", 5.51036e-10, 6.41918e-3, -2.02896e-3, -1.91080e-5, 1.35776e-4, 0.0, 0.0]
```

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+	DDn_Bg   = 31.3970
+	DDp_Bg   = 31.3970
+	DTn_Bg   = 34.3827
+	D3Hep_Bg = 68.7508
+
+	DDn_mr   = 937814
+	DDp_mr   = 937814
+	DTn_mr   = 1124656
+	D3Hep_mr = 1124572
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:25]: Please make this PEP8 compliant

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+	Cs_dict = {"DDn" : DDn_Cs, "DDp" : DDp_Cs, "DTn" : DTn_Cs, "D3Hep" : D3Hep_Cs}
+
+	Bg_dict = {"DDn" : DDn_Bg, "DDp" : DDp_Bg, "DTn" : DTn_Bg, "D3Hep" : D3Hep_Bg}
+	mr_dict = {"DDn" : DDn_mr, "DDp" : DDn_mr, "DTn" : DTn_mr, "D3Hep" : D3Hep_mr}
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:30]: No whitespace before colons (PEP8)

```suggestion
	Cs_dict = {"DDn": DDn_Cs, "DDp": DDp_Cs, "DTn": DTn_Cs, "D3Hep": D3Hep_Cs}

	Bg_dict = {"DDn": DDn_Bg, "DDp": DDp_Bg, "DTn": DTn_Bg, "D3Hep": D3Hep_Bg}
	mr_dict = {"DDn": DDn_mr, "DDp": DDn_mr, "DTn": DTn_mr, "D3Hep": D3Hep_mr}
```

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+	def __init__(self, reaction):
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:32]: Please add docstrings with types of arguments

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+		if reaction in self.known_reactions:
+			self.reaction = reaction
+		else:
+            # THIS IS HORRIBLE PLZ FIX LATER!!!
+			print("Unknown reaction error.")
+			sys.exit()
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:39]:
```suggestion
			raise ValueError(f"Unknown reaction. Should be one of {self.known_reactions}")
```

File: test/test_reactivities_bosch_hale.py
Comment (RemDelaporteMathurin) [inline test/test_reactivities_bosch_hale.py:1]: We can extend this test by testing for different reactions, eventually different temperatures

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+		fraction_bottom = 1 + Temp*(self.C[3] + Temp*(self.C[5] + Temp*self.C[7]))
+
+		theta = Temp/(1 - fraction_top/fraction_bottom)
+		xi    = (self.Bg**2/(4*theta))**(1/3)
+
+		sigmaV = self.C[1]*theta*np.sqrt(xi/(self.mr*Temp**3))*np.exp(-3*xi)
+
+		return sigmaV
+
+	def plot_reactivity(self):
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:57]: Please document this method

Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:57]: On a side note, this method is not tested in the tests.

File: requirements.txt
Comment (RemDelaporteMathurin) [inline requirements.txt:1]: There are a lot of dependencies here (and sub dependencies)

I think our only dependencies here are `numpy` and `matplotlib`

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
Thanks for this contribution! 🎉 🚀 

A few comments on my side.

**Comment by RemDelaporteMathurin** [comment]
We don't know if the tests pass until #3 is addressed

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
+import numpy as np
+import sys
```
Comment (eepeterson) [inline fusion_toolbox/reaction_rate/bosch_hale.py:2]:
```suggestion
```
You can remove this import once you replace the `sys.exit` with an appropriate error

File: test/test_reactivities_bosch_hale.py [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import fusion_toolbox.reaction_rate.bosch_hale as bh
+import numpy.testing as npt
+
+def test_DTn_reactivity():
+    DTn_r= bh.Reactivity("DDp")
+    test_result= 1.3e-17 ## from NRL plasma formulary 
+    test_input = 5.0 # keV
+    npt.assert_almost_equal(DTn_r.get_reactivity(test_input), test_result, decimal=2)
```
Comment (eepeterson) [inline test/test_reactivities_bosch_hale.py:8]:
```suggestion
    assert DTn_r.get_reactivity(test_input) == pytest.approx(test_result) 
```
You can use the pytest.approx function applied to the expected result and assert they are "equal"

File: test/test_reactivities_bosch_hale.py [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import fusion_toolbox.reaction_rate.bosch_hale as bh
+import numpy.testing as npt
+
+def test_DTn_reactivity():
```
Comment (eepeterson) [inline test/test_reactivities_bosch_hale.py:4]: You can use the parametrize decorator we learned about, where `test` would be the temperature in keV and the `expected` would be the reactivity

**Review by eepeterson — CHANGES_REQUESTED** [review]

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    reaction : string
+        String name for the fusion reaction
+
+    """
+    known_reactions = ["DDn", "DDp", "DTn", "D3Hep"]
+
+    _DDn_Cs   = [5.43360e-12, 5.85778e-3, 7.68222e-3, 0.0, -2.96400e-6, 0.0, 0.0]
+    _DDp_Cs   = [5.65718e-12, 3.41267e-3, 1.99167e-3, 0.0, 1.05060e-5, 0.0, 0.0]
+    _DTn_Cs   = [1.1302e-9, 1.51361e-2, 7.51886e-2, 4.60643e-3, 1.35e-2, -1.0675e-4, 1.366e-5]
+    _D3Hep_Cs = [5.51036e-10, 6.41918e-3, -2.02896e-3, -1.91080e-5, 1.35776e-4, 0.0, 0.0]
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:18]: This is still not PEP8 compliant.

You can use a linter to identify the PEP8 issues:

```
python -m pylint [path_to_your_script]
```

**Comment by tmarkj** [comment]
I think I addressed all of the comments in the subsequent commits. Please let me know what you think!

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+    _DDn_Bg = 31.3970
+    _DDp_Bg = 31.3970
+    _DTn_Bg = 34.3827
+    _D3Hep_Bg = 68.7508
+
+    _DDn_mr = 937814
+    _DDp_mr = 937814
+    _DTn_mr = 1124656
+    _D3Hep_mr = 1124572
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:28]: Could you add some comments to explain what Bg, mr, Cs are?

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    _Cs_dict = {"DDn":_DDn_Cs, "DDp":_DDp_Cs, "DTn":_DTn_Cs, "D3Hep":_D3Hep_Cs}
+    _Bg_dict = {"DDn":_DDn_Bg, "DDp":_DDp_Bg, "DTn":_DTn_Bg, "D3Hep":_D3Hep_Bg}
+    _mr_dict = {"DDn":_DDn_mr, "DDp":_DDn_mr, "DTn":_DTn_mr, "D3Hep":_D3Hep_mr}
+
+    def __init__(self, reaction):
+
+        if reaction in self.known_reactions:
+            self.reaction = reaction
+        else:
+            raise ValueError(f"Unknown reaction. Should be one of {self.known_reactions}")
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:39]: Is this tested? 

To test that errors are raised, see the [pytest documentation](https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions)

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+        Returns
+        -------
+        sigmaV : float
+            Calculated Bosch-Hale parameterized in cm^3/s 
+        """
+        C  = self._Cs_dict[self.reaction]
+        Bg = self._Bg_dict[self.reaction]
+        mr = self._mr_dict[self.reaction]
+        fraction_top    = T_i*(C[1] + T_i*(C[3] + T_i*C[5]))
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:58]: PEP8

```suggestion
        fraction_top = T_i*(C[1] + T_i*(C[3] + T_i*C[5]))
```

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        fraction_bottom = 1 + T_i*(C[2] + T_i*(C[4] + T_i*C[6]))
+
+        theta = T_i/(1 - fraction_top/fraction_bottom)
+        xi    = (Bg**2/(4*theta))**(1/3)
```
Comment (RemDelaporteMathurin) [inline fusion_toolbox/reaction_rate/bosch_hale.py:62]: PEP8

```suggestion
        xi = (Bg**2/(4*theta))**(1/3)
```

File: test/test_reactivities_bosch_hale.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
+import numpy.testing as npt
+import pytest
+
+## Testing DTn fusion reaction
+## Tested according to table VIII in H.-S. Bosch and G.M. Hale 1992 Nucl. Fusion 32 611
```
Comment (RemDelaporteMathurin) [inline test/test_reactivities_bosch_hale.py:5]: This should go in the docstrings of the testing function

File: test/test_reactivities_bosch_hale.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+        (10.0, 5.781e-19, 20),
+        (20.0, 2.399e-18, 20),
+    ])
+def test_DDp_reactivity(test, expected, precision):
+    import fusion_toolbox.reaction_rate.bosch_hale as bh
+    DDp_r= bh.Bosch_Hale_reactivity("DDp")
+    npt.assert_almost_equal(DDp_r.get_reactivity(test), expected, decimal=precision) 
+
+## Testing D3Hep fusion reaction
+## Tested according to table VIII in H.-S. Bosch and G.M. Hale 1992 Nucl. Fusion 32 611
```
Comment (RemDelaporteMathurin) [inline test/test_reactivities_bosch_hale.py:53]: Same comment as above

File: test/test_reactivities_bosch_hale.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+@pytest.mark.parametrize(
+    "test, expected, precision",
```
Comment (RemDelaporteMathurin) [inline test/test_reactivities_bosch_hale.py:55]: Why do we need to parametrize the precision here?

File: test/test_reactivities_bosch_hale.py
Comment (RemDelaporteMathurin) [inline test/test_reactivities_bosch_hale.py:1]: I don't think the error raise in the Reactivity class is tested. We could add a test that checks that an error is indeed raised when users provide an invalid reaction string.

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
Some more comments on my side

There are still a few PEP8 non-compliances, don't hesitate to run pylint on the scripts

```
python -m pylint [yourscript]
```

File: test/test_reactivities_bosch_hale.py (hunk 16 above)
Comment (tmarkj) [reply test/test_reactivities_bosch_hale.py:5]: Fixed with https://github.com/tmarkj/fusion-toolbox/commit/11b38930dc50d00fa48acc792eda03745692c0a4

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 19]
```diff
@@ -0,0 +1,67 @@
+import numpy as np
+import matplotlib.pyplot as plt
+
+class Bosch_Hale_reactivity:
```
Comment (eepeterson) [inline fusion_toolbox/reaction_rate/bosch_hale.py:4]:
```suggestion
class BoschHaleReactivity:
```
typically classes are camel case

Comment (tmarkj) [reply fusion_toolbox/reaction_rate/bosch_hale.py]: Fixed with https://github.com/tmarkj/fusion-toolbox/commits/main

File: test/test_reactivities_bosch_hale.py [hunk 20]
```diff
@@ -0,0 +1,118 @@
+import numpy.testing as npt
+import pytest
+
+## Testing DTn fusion reaction
+## Tested according to table VIII in H.-S. Bosch and G.M. Hale 1992 Nucl. Fusion 32 611
```
Comment (RemDelaporteMathurin) [inline test/test_reactivities_bosch_hale.py:5]:
```suggestion
```

File: fusion_toolbox/reaction_rate/bosch_hale.py (hunk 11 above)
Comment (tmarkj) [reply fusion_toolbox/reaction_rate/bosch_hale.py]: Fixed with https://github.com/tmarkj/fusion-toolbox/commit/1b1e5e104cc9fde8318d420baaee8e372f3d3460

File: test/test_reactivities_bosch_hale.py [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+    precision : int 
+        Precision for the comparision between the Bosch-Hale parameterized value
+        and the tabulated value
+    """
+    import fusion_toolbox.reaction_rate.bosch_hale as bh
+    DDp_r= bh.BoschHaleReactivity("DDp")
+    npt.assert_almost_equal(DDp_r.get_reactivity(test), expected, decimal=precision) 
+
+## Testing D3Hep fusion reaction
+## Tested according to table VIII in H.-S. Bosch and G.M. Hale 1992 Nucl. Fusion 32 611
```
Comment (RemDelaporteMathurin) [inline test/test_reactivities_bosch_hale.py:92]:
```suggestion
```

File: test/test_reactivities_bosch_hale.py [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+    precision : int 
+        Precision for the comparision between the Bosch-Hale parameterized value
+        and the tabulated value
+    """
+    import fusion_toolbox.reaction_rate.bosch_hale as bh
+    DDn_r= bh.BoschHaleReactivity("DDn")
+    npt.assert_almost_equal(DDn_r.get_reactivity(test), expected, decimal=precision) 
+
+## Testing DDp fusion reaction
+## Tested according to table VIII in H.-S. Bosch and G.M. Hale 1992 Nucl. Fusion 32 611
```
Comment (RemDelaporteMathurin) [inline test/test_reactivities_bosch_hale.py:63]:
```suggestion
```

File: test/test_reactivities_bosch_hale.py [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+    precision : int 
+        Precision for the comparision between the Bosch-Hale parameterized value
+        and the tabulated value
+    """
+    import fusion_toolbox.reaction_rate.bosch_hale as bh
+    DTn_r= bh.BoschHaleReactivity("DTn")
+    npt.assert_almost_equal(DTn_r.get_reactivity(test), expected, decimal=precision) 
+
+## Testing DDn fusion reaction
+## Tested according to table VIII in H.-S. Bosch and G.M. Hale 1992 Nucl. Fusion 32 611
```
Comment (RemDelaporteMathurin) [inline test/test_reactivities_bosch_hale.py:34]:
```suggestion
```

**Review by RemDelaporteMathurin — APPROVED** [review]
Just a few comments to remove some dead code.
Please open an issue to add a test for the error raise later.

Then we are ready to merge this!

File: test/test_reactivities_bosch_hale.py (hunk 18 above)
Comment (tmarkj) [reply test/test_reactivities_bosch_hale.py:86]: The Bosch Hale reactivities are just parameterizations fit to the fusion reactivity. So the values it gives will not exactly match the experimental or integrated reactivities. The precision is needed to make the tests pass when the values are close enough.

File: fusion_toolbox/reaction_rate/bosch_hale.py (hunk 15 above)
Comment (tmarkj) [reply fusion_toolbox/reaction_rate/bosch_hale.py]: Fixed with https://github.com/tmarkj/fusion-toolbox/commit/1b1e5e104cc9fde8318d420baaee8e372f3d3460

File: fusion_toolbox/reaction_rate/bosch_hale.py (hunk 12 above)
Comment (tmarkj) [reply fusion_toolbox/reaction_rate/bosch_hale.py]: Added an explanation with commit: 2ab912ed44fd9dfa54be47bfc691a56cf2106a45

**Comment by RemDelaporteMathurin** [comment]
@tmarkj you have some conflicts now that #5 has been merged. Can you resolve them?

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 24]
```diff
@@ -0,0 +1,71 @@
+import numpy as np
+
```
Comment (eepeterson) [inline fusion_toolbox/reaction_rate/bosch_hale.py:2]: there should be 2 line spacing for module level classes and functions and single line spacing within functions and classes per PEP8

File: fusion_toolbox/reaction_rate/bosch_hale.py [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+    def __init__(self, reaction):
+
+        if reaction in self.known_reactions:
+            self.reaction = reaction
+        else:
+            raise ValueError(f"Unknown reaction. Should be one of {self.known_reactions}")
+
+
+    def get_reactivity(self, T_i):
```
Comment (eepeterson) [inline fusion_toolbox/reaction_rate/bosch_hale.py:47]:
```suggestion
    def get_reactivity(self, ion_temperature):
```
Something more descriptive following lower case conventions might be better

**Review by eepeterson — CHANGES_REQUESTED** [review]
Generally I would move your bosch_hale.py file up one level to something called `reactivities.py` in the case that you wanted to add a different class based on a different papers, it would just go here. Then add the line `from fusion_toolbox.reactivities import *` to `__init__.py`

File: test/test_reactivities_bosch_hale.py (hunk 9 above)
Comment (tmarkj) [reply test/test_reactivities_bosch_hale.py]: Fixed this with 268d8135ec53e82305ac64b1cfea7851caf2b443

File: fusion_toolbox/reaction_rate/bosch_hale.py (hunk 24 above)
Comment (tmarkj) [reply fusion_toolbox/reaction_rate/bosch_hale.py:2]: Fixed with 6e8181d20090b5ee6cf75efc2ee198de7d8cc5f7

**Comment by tmarkj** [comment]
> Generally I would move your bosch_hale.py file up one level to something called `reactivities.py` in the case that you wanted to add a different class based on a different papers, it would just go here. Then add the line `from fusion_toolbox.reactivities import *` to `__init__.py`

Fixed this with f1f6a5fa5b554b86518df71f26a608076acd0552

File: fusion_toolbox/reactivity.py [hunk 26]
```diff
@@ -0,0 +1,71 @@
+import numpy as np
+
+
+class BoschHaleReactivity:
+    """Class for Bosch Hale reactivity (H.-S. Bosch and G.M. Hale 1992 Nucl. Fusion 32 611)
+
+    Parameters
+    ----------
+    reaction : string
```
[the remark this answers is not in the document]
Comment (tmarkj) [reply fusion_toolbox/reactivity.py:9]: Fixed with 65353c1161b09805bf335d3b7bd1caf0e36b604c

**Review by eepeterson — APPROVED** [review]
nicely done!

**Pull request merged**